Address HTTPDIR review feedback and add structured partition keys (#166)
ietf-wg-httpapi/ratelimit-headers: draft-ietf-httpapi-ratelimit-headers.md

--- middle

# Introduction

Rate limiting of HTTP {{HTTP}} clients is a widespread practice. Servers use quota mechanisms to avoid overload, to ensure an equitable distribution of computational resources, or to enforce other policies such as monetization. Typically, servers limit the number of acceptable requests in a given time window (e.g. 10 requests per second). When a quota is exceeded, servers usually respond with a 4xx HTTP status code (e.g. 429 or 403) or adopt more aggressive policies like dropping connections.

Quotas may be enforced on different bases (e.g. per user, per IP, per geographic area) and at different levels. To help clients pace their requests, servers can expose quota information via HTTP header fields. This document defines a set of standard HTTP header fields to enable servers to communicate rate limiting information to clients:

- RateLimit-Policy: quota policies defined by the server. Client requests consume quota.
- RateLimit: the quota currently available under specific policies at the time the response is generated.

> These fields enable servers to communicate complex rate limiting policies, including using multiple and variable time windows and dynamic quotas, and implementing concurrency limits. This information helps improve the resiliency of HTTP infrastructure by providing clients with the information needed to throttle their requests and avoid 4xx or 5xx responses.

Reviewer: ```suggestion
These fields enable servers to communicate complex rate limiting policies, including using multiple and variable time windows and dynamic quotas, and implementing concurrency limits. This information helps improve the resiliency of HTTP infrastructures by providing clients with the information needed to throttle their requests and avoid 4xx or 5xx responses.
```

  Service Level Agreement:
  : Conveyed quota hints do not imply any service guarantee. Server is free to throttle clients that adhere to the server’s recommended limits under certain circumstances.

## Conventions and Terminology

{::boilerplate bcp14}

This document uses the terms List, Item, Integer, String, Token, Boolean, Parameters, and Byte Sequence from {{Section 3 of STRUCTURED-FIELDS}} to specify syntax and parsing, along with the concept of "bare item".

The term "problem type" in this document is to be interpreted as described in [PROBLEM].

  Quota:
  : A quota is an allocation of capacity used by a server to limit client requests. That capacity is measured in quota units that the client may consume within a time window.

  Quota Unit:

>   : A quota unit is the unit of measurement used to measure the activity of a client.

Reviewer: ```suggestion
  : A quota unit is the unit of measurement for the activity of a client.
```

  : A time window indicates a period of time associated to the allocated quota.

  Quota Policy:
  : A quota policy is implemented by a server to regulate activity within a specified quota partition, quantified in quota units, over a defined time window. This activity is restricted to a predefined limit, known as the quota. Servers MAY advertise quota policies. More than one quota policy can affect a given request.

  Service Limit:
  : A service limit is the currently available quota under a specific quota policy and, if defined, the effective time window within which the client can use no more than the available quota.

# RateLimit-Policy Field {#ratelimit-policy-field}

The "RateLimit-Policy" response header field is a List of Strings. Each list item identifies a Quota Policy ({{quotapolicy-item}}).

This field MUST NOT appear in a trailer section.

> The "RateLimit-Policy" field enables clients to pace their requests based on policy information provided by the server. The field value SHOULD remain consistent over a sequence of HTTP responses, as policy information is not expected to change frequently. Situations where throttling constraints are highly dynamic are better served using the RateLimit field ({{ratelimit-field}}) that communicates the latest service information a client can react to. Both fields can be communicated by the server when appropriate.

Reviewer: 1. explain the reason for consistency :)

```suggestion
The "RateLimit-Policy" field enables clients to pace their requests based on policy information provided by the server. The field value SHOULD remain consistent over a sequence of HTTP responses, changes would prevent clients from correctly following the communicated policy. Situations where throttling constraints are highly dynamic are better served using the RateLimit field ({{ratelimit-field}}) that communicates the latest service information a client can react to. Both fields can be communicated by the server when appropriate.
```

A server might also declare a policy that applies only to a specific method:

~~~
   RateLimit-Partition: "reads";user_id;method=GET
~~~

A client receiving this knows the "reads" policy only applies to GET requests. For non-GET requests, this policy is not relevant and can be ignored.

A client that knows its own values for the declared dimensions can construct the expected partition key for a future request and look it up in its local quota tracking state, enabling proactive traffic shaping without requiring API-specific knowledge.

If a server provides a "pk" parameter without a corresponding RateLimit-Partition field item, the structure of the partition key is not defined by this specification and any interpretation of its structure MUST be communicated out of band. Clients MAY use heuristics to estimate partition membership based on observed partition key values and request similarity.

# Server Behavior {#providing-ratelimit-fields}

A server MAY return RateLimit header fields independently of the response status code. This includes throttled responses. This document does not mandate any correlation between the RateLimit header field values and the returned status code.

> Servers SHOULD avoid returning RateLimit header fields in redirection responses (i.e., responses with 3xx status codes) because a low available quota could prevent the client from following the redirect. For example, given the RateLimit header fields below, a client could decide to wait 10 seconds before following the "Location" header field (see {{Section 10.2.2 of HTTP}}), because the available quota is 0.

Reviewer: Maybe going full normative.

```suggestion
Servers SHOULD NOT return RateLimit header fields in redirection responses (i.e., responses with 3xx status codes) because a low available quota could prevent the client from following the redirect. For example, given the RateLimit header fields below, a client could decide to wait 10 seconds before following the "Location" header field (see {{Section 10.2.2 of HTTP}}), because the available quota is 0.
```

Author: I don’t think changing this to  SHOULD NOT  addresses the underlying problem. The risk is that a client might incorrectly apply quota information for the original request to the request identified by  Location .

RateLimit fields on a 3xx response can still provide useful information about the original request. Instead, we should clarify their scope and state:

> Clients MUST NOT assume that RateLimit header fields in a redirection response apply to the request identified by the Location field.
